Improve constrained optimization

till-m opened the pull request:
- Fix #363: ScreenLogger now prints "allowed"-ness
- Assert that `lb` (lower bound) is lower than `ub` upper bound
    Something that I mentioned in the original PR but then forgot to do.
- Merge `ConstrainedTargetSpace` into `TargetSpace`
Originally I thought there would be less clutter if I subclass the original `TargetSpace`, but I think it will make future maintenance much harder since code will need to be changed in multiple places at once (see e.g. #365).
- Fix a bug related to one-dimensional constraints with bounds of type `int` (presumably the bug mentioned in #361)
- Allow loading of logs with constraint information (see #363 as well)
- Mention the constrained optimization notebook in the quickstart section of the README.
- Added a stopgap to deal with registering a duplicate point when loading logs.

bwheelz36 commented:
I have a feeling that test may have failed randomly, can you make a tiny change and push to force it to run again?
This failure would be fixed when I address #353

till-m commented:
I added a stopgap so that I can make sure loading from logs behaves correctly and replaced the (imho) slightly obscure `KeyError` that was raised. 

@bwheelz36 feel free to review at your convenience -- I've addressed all the things I wanted to address.
